### ISSUE TITLE
[RFC] RPC BENCHMARK

### ISSUE THREAD

**Issue by osandoval-fb** [pytorch/pytorch#43561]
## Background

This will allow potential users to see what kind of throughput they can expect if they use RPC for their use case.  Also so the throughput effects of changes made to RPC can be measured.

Users of the benchmark can invoke a script, passing multiple arguments to specify what kind of environment they want to see benchmark numbers for. 

## API

The following arguments will be available
- `model_layers` (default=2)
- `model_weight` (default=10)
- `model_bias` (default=10)
- `agent_count` (default=1)
- `observer_count` (default=100)
- `batch` (default=True)
- `rpc_framework` (default=ProcessGroup), (options=ProcessGroup,TensorPipe,Postman)
- `device_type` (default=CPU), (options=CPU,GPU)

##Example Output

```
-----------------------------------
PyTorch RPC benchmark suite
-----------------------------------

* PyTorch version: 1.4.0a0+05140f0
* CUDA version: 10.0
* Core Count: 12
* Device: CPU
* Batch: True
* Observers: 100
* Agents: 1
* Framework: ProcessGroup
* Model Layers, Weight, Bias: (2, 10, 10)

            sec/iter    ex/sec      sec/iter    ex/sec      sec/iter    ex/sec      sec/iter    ex/sec
  p50:  0.162s     197/s  p75:  0.162s     197/s  p90:  0.162s     197/s  p95:  0.162s     197/s
```

## Implementation Plan
TBD

**Comment by mrshenli** [comment]
Hey @osandoval-fb, thanks for posting this. You might want to mention that the first benchmark will be RL-specific, and we will need to explain the meaning of each argument and how they affect the benchmark behavior. BTW, what is `model_weight` and `model_bias` parameter, do you mean the `in_features` and `out_features` arg in the `nn.Linear` module?

I am trying to expand your RFC to the following. The design section is just one way to implement this. Please feel free to propose other solutions. 


--------------------


## Overview

Build an RPC benchmark to mimic multi-observer single-agent applications under different configurations, measuring throughput and latency. This benchmark will spawn multiple RPC worker processes, where one of them will be the agent and the rest will be observers. Every observer only talks to the agent. More specifically, in each iteration, it generates an observer state (a Tensor) and then send it to the agent. Upon receiving the state, the agent runs it through its policy (an `nn.Module`) to generate an action (a Tensor), and sends the action back to the observer. The observer applies the action to its own environment and gets a reward and the next state. An example can be found in [this tutorial](https://pytorch.org/tutorials/intermediate/rpc_async_execution.html#batch-processing-cartpole-solver).


## Design

### Concepts

There are four concepts in this benchmark:

* **Launcher Script** parses argument and launches all other processes in the current machine accordingly. Users need to run the launcher script on each machine.
* **Coordinator Process** create all agent and observers objects using `rpc.remote`, shares the `RRef` objects when necessary, and collect performance numbers. There is only one Coordinator globally.
* **Agent Process** hosts the policy `nn.Module`, listens for environment states from observers, runs the states through its policy, and responds with the action.
* **Observer Process** owns its dedicated environment, extracts states from its environment, sends the states to the agent and gets an action as a return value.

Among the four concepts, the Launcher and the Coordinator should be the same for all different configurations. The Agent and the Observer can be custom implementations. 

As there are only one Coordinator and only one Agent globally, there rank is fixed to be 0 and 1.

### Arguments

Launch arguments: (some of the names are borrowed from [DDP launcher script](https://github.com/pytorch/pytorch/blob/master/torch/distributed/launch.py))

* `--world_size`: [default `3`.] the total number of processes, including the Coordinator process and the Agent process, i.e., this is number of observers + 2.
* `--nnode`: [default `1`.] the number of machines running this benchmark.
* `--node_rank`: [default `0`.] The rank of the current node for multi-node distributed benchmark.
* `--nproc`: [default `-1`.] Number of processes in the current node. This does not need to be the same across all nodes. If not given, this will be the same as `--world_size`.
* `--master_addr`: [default `"127.0.0.1"`] IP address of rendezvous master. See `rpc_backend_options` in `init_rpc`.
* `--master_port`: [default `29501`] Port number of rendezvous master. See `rpc_backend_options` in `init_rpc`.
* `--rpc_backend`: [default `"TENSORPIPE"`] RPC backend type. See [RPC backends](https://pytorch.org/docs/master/rpc.html#backends).

Coordinator arguments:

* `--observer`: default `"dummy"`. Observer type, which corresponds to a specfic observer implementation. We will provide a dummy implementation. Users can also customize their observers.
* `--agent`: default "dummy". Agent type, which corresponds to a specific agent implementation. Users can customize this as well. 
* `--device`: default "cpu". Controls whether the states and policy lives on CPU or GPU. For "CUDA", states and policy will be placed to the default CUDA device. 
* `--batch`: default `True`. Controls whether the observer and agent uses batch RPC or not. See [`async_execution`](https://pytorch.org/docs/master/rpc.html#torch.distributed.rpc.functions.async_execution).

Dummy Agent arguments:

* `--nlayers`: default `0`. The dummy agent's policy is an `nn.Sequential` of `nn.Linear` modules. The argument controls the number of linear layers, which can be interpreted as the complexity of the policy.
* `--in_features`: default `10`. The `in_features` argument for `nn.Linear`.
* `--out_features`: default `10`. The `out_features` argument for `nn.Linear`.

Dummy Observer arguments:

* `--state_size`: default 0. This is a string of comma-separated dimensions, e.g., `--state_size=10,20,10` will result in states of `torch.zeros(10, 20, 10)`.

### Run the Benchmark

**Scenario 1**: run the benchmark on a single machine to measure the throughput and latency of 20 observers:

```bash
$ python benchmarks/distributed/rpc/launch.py --world_size=22 --nlayers=10 --in_features=20 --state_size=50,20

   Agent throughput --  p50:  0.097s     329/s  p75:  0.097s     329/s  p90:  0.097s     329/s  p95:  0.097s     329/s
Observer throughput --  p50:  0.097s     329/s  p75:  0.097s     329/s  p90:  0.097s     329/s  p95:  0.097s     329/s
      Agent latency --  p50:  0.097s     329/s  p75:  0.097s     329/s  p90:  0.097s     329/s  p95:  0.097s     329/s
   Observer latency --  p50:  0.097s     329/s  p75:  0.097s     329/s  p90:  0.097s     329/s  p95:  0.097s     329/s
```

**Scenario 2**: run the benchmark on two machines, where one machine is dedicated to agent

```bash
# on machine 1
$ python benchmarks/distributed/rpc/launch.py --world_size=22 --nproc=2 --nnode=2 --nlayers=10 --in_features=20 --state_size=50,20

   Agent throughput --  p50:  0.097s     329/s  p75:  0.097s     329/s  p90:  0.097s     329/s  p95:  0.097s     329/s
Observer throughput --  p50:  0.097s     329/s  p75:  0.097s     329/s  p90:  0.097s     329/s  p95:  0.097s     329/s
      Agent latency --  p50:  0.097s     329/s  p75:  0.097s     329/s  p90:  0.097s     329/s  p95:  0.097s     329/s
   Observer latency --  p50:  0.097s     329/s  p75:  0.097s     329/s  p90:  0.097s     329/s  p95:  0.097s     329/s
```

```bash
# on machine 2
$ python benchmarks/distributed/rpc/launch.py --world_size=22 --nproc=20 --nnode=2 --node_rank=1
```

### Customize Agent and Observer


As we are going to support custom observer and agent implementations, we need a base class to specify the APIs that a custom type need to implement. Below is an example, we can materialize more details as we implement this. 

```python
class AgentBase:
  def select_action(self, observer_id, state):
    pass


  def finish_episode(self):
    # can train and update policy here
    pass

```

```python
class ObserverBase:
  def reset(self):
    # reset env and return the initial state
    pass

  def step(self, action):
    # apply action and return new state and rewards
    pass

```
